designs/2023-rule-performance-dashboard/README.md

```diff
@@ -0,0 +1,276 @@
+```json
+{
+    "stats": {
+        "directives": 0,
+        "violations": 0,
+        "fixPasses": 0,
+        "timing": {
+            "rules": {
+                "total": 1.650541,
+                "no-regex-spaces": {
+                    "total": 1.650541,
+                    "create": 0.0435,
+                    "Literal": 0.030583,
+                    "CallExpression": 0,
+                    "NewExpression": 1.576458
+                }
+            },
+            "fix": 0,
+            "parse": 1.911792,
+            "total": 3.5623329999999997
+        }
+    }
+}
+```
+
+A special [formatter](https://eslint.org/docs/latest/use/formatters/) `html-rule-performance` enables easy ingestion and interpretation of this data in form of a dashboard.
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+An analysis of rule performance in ESLint can already be carried out by setting the [TIMING](https://eslint.org/docs/latest/extend/custom-rules#profile-rule-performance) environment variable. However, **additional scripting is still required** to collect/extract more *granular* timing data (lint time per file per rule) by following one of the two approaches:
+
+1. Running ESLint per rule, per file and then collecting the file/rule time data output:
+
+    ```bash
+    TIMING=1 DEBUG=eslint:cli-engine eslint --no-eslintrc --rule ... file
+    ```
+
+2. Doing a single ESLint run on all files and then extracting the file/rule time data output:
+
+    ```bash
+    TIMING=all DEBUG=eslint:cli-engine eslint ... files
+    ```
+
+In addition, one needs to create an overview for an effective presentation of the results.
+
+Since ESLint already collects most of this data internally, it would be more *convenient and shareable* to have:
+
+   1. The **timing data** from ESLint exposed to the formatters.
+   2. The possibility to also gain more insight into [other runtime statistics](https://github.com/eslint/eslint/issues/14597#issuecomment-1003863524) (not time related).
+   3. A **built-in formatter** for a dashboard of that data.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+------
+
+1. **Adding the new `stats` option to ESLint:**
+
+ - [_docs/src/use/command-line-interface.md_](https://github.com/mnkiefer/eslint/pull/1/files#diff-80937556f9a8e68352718255c20699fce061fa72760504c6accc3fc6b3b0613aR117): First, we document its *name* and *purpose* for the [Command Line Interface Reference](https://eslint.org/docs/latest/use/command-line-interface) under the *Miscellaneous* category.
+     ```md
+     --stats     Add additional statistics to lint results
```

Comment:
What what does the CLI output of `eslint --stats` look like? In my comment https://github.com/eslint/eslint/issues/14597#issuecomment-1003863524, I drew out some human-readable tables that provide timing information like the `TIMING` environment variable does but with the additional metrics.

My original expectation for this feature was that all the metrics/timing would be available in these places:

1. Programmatically via the Node API for anyone wanting to build formatters or write scripts on top of the data.
2. Human-readable CLI output for ESLint users to quickly view the data on command line, like the `TIMING` environment variable. 
3. Via any first-party or third-party formatters for users wanting additional visualizations of the data.

Reply:
Indeed, this is still an open question that needs to be addressed. 👍

@nzakas, please advise.

Reply:
Related comment thread where I propose unifying `TIMING` and `stats`: https://github.com/eslint/rfcs/pull/108#discussion_r1221725356

Reply:
The idea behind `--stats` is ONLY to add additional data into the result object and it shouldn't affect the output of ESLint in any way. This will allow the creation of custom formatters that can take the stats and format them however they'd like.

As such, I disagree that we should merge `TIMING` and `stats`. These are two different (although related) use cases.

I believe this RFC will allow third-party developers to create custom formatters that are better than the default of what we show with `TIMING` right now. We want to focus on getting them the data they need to do so.

Reply:
I fully agree (as mentioned in my earlier comment https://github.com/eslint/rfcs/pull/108#discussion_r1218237114).